40-tpm2-pre-boot-event-log-support.md

```diff
@@ -0,0 +1,277 @@
+is different as it encodes the MAC address of the NIC). Our experiments with the 
+"TPM 2.0 Pre-Boot Event log" demonstrated that, with a "recent enough" Kernel (e.g. 5.4.X), 
+and a "recent enough" version of tpm2-tools, we can access the contents of the aforementioned 
+log to re-create (or debug) the values of PCRs [0-9],14 without requiring a 
+pre-measurement (i.e., a "golden value").
+
+### Goals
+
+<!--
+List the specific goals of the enhancement.  What is it trying to achieve?  How will we
+know that this has succeeded?
+-->
+
+- Add support for the `agents` to (conditionally) read and ship the TPM 2.0 event log
+- Add support on the `verifier` to replay the event log and reconstruct the values of PCR [0-9],14
+- Add support on the `tenant` CLI to specify an "event log desired state", which can then be 
+consumed by the `verifier` to attest a given event log
+
+### Non-Goals
+
+<!--
+What is out of scope for this enhancement?  Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+- The "policy", which will take a given event log, a given "event log desired state" 
+and issue a pass/fail on the former based on the latter is outside of the scope of this
+enhacement. We will provide a way for a `verifier` to execute a generic policy, but the
+actual policy code will be provided by Keylime deployers/users.
+
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation.  The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+- The `tenant` CLI will provide a new option to indicate an "event log desired state"
+- The `agent`, upon being notified that a new "event log desired state" exists,
+will read the TPM2.0 pre-boot event log from /sys/kernel/security/tpm0/binary_bios_measurements
+and ship it back to the `verifier`
+- Whenever the `agent` reads the TPM2.0 pre-boot event log, it will also
+read (and ship to the `verifier`) the contents of PCRs [0-9], 14.
+- The verifier will be extended to support the replaying of the event log, 
+making sure it matches the contents of the aforementioned PCRs.
+- The verifier will also be extended to support the execution of generic
+python code, in the form of a "policy", which will read the contents of 
+the TPM2.0 Pre-Boot Event Log extracted by the agent, the contents of the 
+"event log desired state" (supplied via `tenant` CLI and stored on the database) 
+and then apply a pass/fail attestation test to the event log.
+- Test cases will be written that a very simple "policy", which will simply
+read the contents of the event log and attest it as legitimate, will be available
+by default. This means that the TPM2.0 Pre-Boot Event Log will be simply read on its
+entirety and that is can properly reconstruct PCRs [0-9],14. 
+
```

Comment:
One more deliverable I would like to have is documentation about this feature (can be mostly based on this document) and sample policy code so users can easily see what data structures and arguments are involved in how the code is invoked.